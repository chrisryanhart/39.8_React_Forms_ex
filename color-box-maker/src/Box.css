.Box {
    /* height: 100px;
    width: 100px;
    background-color: greenyellow; */
    border: solid black 2px;
}