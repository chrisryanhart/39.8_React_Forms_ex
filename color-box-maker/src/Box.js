import React from 'react';
import './Box.css';

const Box = ({id, remove, width="100px", height='100px', backgroundColor='greenyellow'}) => {
    // have to remove the box upon click

    const handleClick = () => remove(id);
    
    return (
        <>
           <div className='Box' key={id} style={{width, height, backgroundColor}}/>
           <button onClick={handleClick}>X</button>
        </>
    );
}

export default Box;