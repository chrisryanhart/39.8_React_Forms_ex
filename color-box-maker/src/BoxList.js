import React, {useState} from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from "uuid";

const BoxList = () => {

    const INITIAL_STATE = [
            {id: uuid(), width: "100px", height: "100px", backgroundColor: "blue"},
            {id: uuid(), width: "100px", height: "100px", backgroundColor: "yellow"}
        ]
        
    const [boxes,setBoxes] = useState(INITIAL_STATE);

    const addBox = (width,height,backgroundColor) => {
        setBoxes(data => (
             [...data,
              {id: uuid(), width,height,backgroundColor}
            ]))
    }

    const removeBox = (id) => {
        setBoxes(data => (
            data.filter(ele => ele.id !== id)
        ));
    }

    return (
        <>
            {boxes.map(box => (
                <Box 
                    key={box.id}
                    id={box.id}
                    width={box.width} 
                    height={box.height} 
                    backgroundColor={box.backgroundColor}
                    remove={removeBox}
                />))}
           <NewBoxForm addBox={addBox}/>
        </>
    );
}

// addBox={addBox} value={boxFormData}

export default BoxList;