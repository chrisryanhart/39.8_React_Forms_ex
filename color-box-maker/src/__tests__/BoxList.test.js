import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import BoxList from '../BoxList';
import Box from '../Box';

it("renders without crashing", function() {
    render(<BoxList/>);
})

it("renders without crashing", function() {
    const {asFragment} = render(<BoxList/>);
    expect(asFragment()).toMatchSnapshot();
})

// add new box

it("can add a new Box", function() {
    const { getAllByText, getByLabelText, queryByText } = render(<BoxList/>);

    expect(getAllByText('X')).toHaveLength(2);

    expect(getAllByText('X')[1].previousSibling).toHaveStyle(`
    width: 100px;
    height: 100px;
    background-color: yellow;`);

    const colorInput = getByLabelText('BG Color:');
    const widthInput = getByLabelText('Width:');
    const heightInput = getByLabelText('Height:');
    const submitBtn = queryByText('Add Box');

    fireEvent.change(colorInput, {target: {value: "red"}});
    fireEvent.change(widthInput, {target: {value: "200px"}});
    fireEvent.change(heightInput, {target: {value: "200px"}});

    fireEvent.click(submitBtn);

    expect(getAllByText('X')).toHaveLength(3);
})

it("can delete a Box", function() {
    const { getAllByText, getByLabelText, queryByText } = render(<BoxList/>);

    expect(getAllByText('X')).toHaveLength(2);

    const deleteBtn = getAllByText('X')[0];

    fireEvent.click(deleteBtn);

    expect(getAllByText('X')).toHaveLength(1);
})