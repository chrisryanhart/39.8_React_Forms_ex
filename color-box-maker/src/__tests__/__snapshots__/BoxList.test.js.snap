// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="Box"
    style="width: 100px; height: 100px; background-color: blue;"
  />
  <button>
    X
  </button>
  <div
    class="Box"
    style="width: 100px; height: 100px; background-color: yellow;"
  />
  <button>
    X
  </button>
  <form>
    <div>
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        value=""
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        value=""
      />
      <label
        for="backgroundColor"
      >
        BG Color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        value=""
      />
    </div>
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
